llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java

```diff
@@ -0,0 +1,114 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.llap.daemon.impl;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.service.AbstractService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * LocalDirCleaner is an LLAP daemon service to clean up old local files. Under normal circumstances,
+ * intermediate/local files are cleaned up (typically at end of the DAG), but daemons crash sometimes,
+ * and the attached local disk might end up being the same when a new daemon starts (this applies to
+ * on-prem as well as cloud scenarios).
+ */
+public class LocalDirCleaner extends AbstractService {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalDirCleaner.class);
+
+  private String[] localDirs;
+
+  private long cleanupIntervalSec;
+  private long fileModifyTimeThresholdSec;
+
+  ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+  public LocalDirCleaner(String[] localDirs, Configuration conf) {
+    super("LocalDirCleaner");
+    this.localDirs = localDirs;
+    this.cleanupIntervalSec = getInterval(conf);
+    this.fileModifyTimeThresholdSec = getFileModifyTimeThreshold(conf);
+    LOG.info("Initialized local dir cleaner: interval: {}s, threshold: {}s", cleanupIntervalSec,
+        fileModifyTimeThresholdSec);
+  }
+
+  @Override
+  public void serviceStart() throws IOException {
+    scheduler.scheduleAtFixedRate(this::cleanup, 0, cleanupIntervalSec, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void serviceStop() throws IOException {
+    // we can shutdown this service now and ignore leftovers, because under normal circumstances,
+    // files from the local dirs are cleaned up (so LocalDirCleaner is a best effort utility)
+    scheduler.shutdownNow();
+  }
+
+  private long getFileModifyTimeThreshold(Configuration conf) {
+    return HiveConf.getTimeVar(conf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_FILE_MODIFY_TIME_THRESHOLD,
+        TimeUnit.SECONDS);
+  }
+
+  private long getInterval(Configuration conf) {
+    return HiveConf.getTimeVar(conf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_CLEANUP_INTERVAL, TimeUnit.SECONDS);
+  }
+
+  private void cleanup() {
+    Instant deleteBefore = Instant.now().minus(fileModifyTimeThresholdSec, ChronoUnit.SECONDS);
+
+    for (String localDir : localDirs) {
+      Path pathLocalDir = Paths.get(localDir);
+      cleanupPath(deleteBefore, pathLocalDir);
+    }
```

Comment:
Can be simplified by stream api:
```suggestion
localDirs.forEach(localDir -> cleanupPath(deleteBefore, Paths.get(localDir)));
```

Reply:
good catch, I'll change